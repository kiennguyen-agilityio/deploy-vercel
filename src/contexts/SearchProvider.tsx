import { createContext, useState, useMemo, ReactNode } from 'react';

export interface SearchContextProps {
  searchValue: string;
  onSearchValue: (value: string) => void;
  onResetSearchValue: () => void;
}
interface SearchProviderProps {
  children: ReactNode;
}

export const SearchContext = createContext<SearchContextProps>({} as SearchContextProps);

export const SearchProvider: React.FC<SearchProviderProps> = ({ children }) => {
  const [searchValue, setSearchValue] = useState<string>('');

  const resetSearchValue = (): void => {
    setSearchValue('');
  };

  const setValue = (value: string): void => {
    setSearchValue(value);
  };

  const contextValue = useMemo(
    () => ({
      searchValue,
      setSearchValue,
      onResetSearchValue: resetSearchValue,
      onSearchValue: setValue,
    }),
    [searchValue, setSearchValue]
  );

  return <SearchContext.Provider value={contextValue}>{children}</SearchContext.Provider>;
};
