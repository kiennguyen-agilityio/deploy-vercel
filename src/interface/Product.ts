interface Product {
  id: string;
  name: string;
  sale: string;
  price: number;
  discountPercent: number;
  image: string;
  description: string;
  quantity: number;
  color: string;
  subcategory: string;
}

export default Product;
