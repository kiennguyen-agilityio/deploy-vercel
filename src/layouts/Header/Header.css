.navbar {
  display: flex;
  justify-content: center;
  align-items: center;
  width: auto;
  height: 120px;
  background: var(--color-white);
}

.logo {
  padding-right: 50px;
}

.fa-sharp {
  font-size: var(--font-size-medium);
  color: var(--color-black);
}

.shop-name {
  margin-left: 15px;
  color: var(--color-black);
  font-family: 'Open Sans', sans-serif;
  font-weight: var(--font-weight-extrabold);
}

.search-bar-container {
  position: relative;
  display: inline-block;
  width: 627px;
  height: 60px;
  border: 1px solid var(--color-black);
  border-radius: 10px;
}

.search-bar {
  font-family: 'Open Sans', sans-serif;
  box-sizing: border-box;
  width: 100%;
  height: 100%;
  padding-left: 25px;
  padding-right: 25px;
  border: none;
  background-color: transparent;
}

.icons {
  display: flex;
  align-items: center;
}

.icon {
  margin-left: 50px;
  color: var(--color-black);
}

.button-container {
  display: flex;
  margin-left: 50px;
  align-items: center;
  justify-content: center;
}

.btn-sign-up {
  border: none;
  background: none;
  width: 90px;
  height: 27px;
  font: normal var(--font-weight-semibold) var(--font-size-sm) / var(--line-height-md) 'Open Sans';
  color: var(--color-black);
}

.btn-login {
  width: 123px;
  height: 60px;
  background: var(--color-background-button);
  border-radius: 10px;
  font: normal var(--font-weight-semibold) var(--font-size-sm) / var(--line-height-md) 'Open Sans';
  margin-left: 30px;
  border: none;
}

.cart-icon-container {
  position: relative;
}

.cart-item-count {
  position: absolute;
  top: -10px;
  right: -10px;
  background-color: var(--color-secondary);
  color: var(--color-white);
  border-radius: 50%;
  padding: 4px;
  font-size: var(--font-size-12);
}
