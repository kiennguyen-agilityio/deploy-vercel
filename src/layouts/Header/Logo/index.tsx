import { memo } from 'react';
import { Link } from 'react-router-dom';

const Logo = () => (
  <Link to="/">
    <div className="logo">
      <i className="fa-sharp fa-solid fa-bag-shopping" />
      <span className="shop-name">Shop Bags</span>
    </div>
  </Link>
);

export default memo(Logo);
