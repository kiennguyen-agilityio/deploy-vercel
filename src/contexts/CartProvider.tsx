import { createContext, useMemo, useCallback, ReactNode, useReducer, useContext } from 'react';
import { mutate } from 'swr';

// constants
import {
  ADD_PRODUCT_FAILURE,
  DELETE_PRODUCT_FAILURE,
  UPDATE_PRODUCT_FAILURE,
  ADD_PRODUCT_REQUEST,
  ADD_PRODUCT_SUCCESS,
  DELETE_PRODUCT_REQUEST,
  DELETE_PRODUCT_SUCCESS,
  UPDATE_PRODUCT_REQUEST,
  UPDATE_PRODUCT_SUCCESS,
} from '@/constants/actionType';

// services
import { addToCart, deleteCartItem, updateItem } from '@/services/api-action';

// hooks
import { useCartProducts } from '@/components/hooks/fetch';

// constants
import { CART_API } from '@/constants/api';

// reducers
import cartReducer, { ProductCartState } from '@/reducers/cartReducer';

// interfaces
import Product from '@/interface/Product';

export const initialState: ProductCartState = {
  error: '',
  isLoading: false,
  isAdding: false,
  products: [],
};

export interface CartContextProps {
  error: string;
  isAdding: boolean;
  isLoading: boolean;
  products?: Product[];
  onAddToCart: (product: Product) => void;
  onUpdateCart: (productId: string, updatedData: Partial<Product>, url: string) => Promise<void>;
  onDeleteProductFromCart: (productId: string) => void;
}

export const CartContext = createContext<CartContextProps>({
  error: '',
  isAdding: false,
  isLoading: false,
  products: [],
  onAddToCart: () => undefined,
  onUpdateCart: async () => {},
  onDeleteProductFromCart: async () => {},
});

interface CartProviderProps {
  children: ReactNode;
}

export const useProductCartContext = () => useContext(CartContext);

const CartProvider: React.FC<CartProviderProps> = ({ children }) => {
  const [state, dispatch] = useReducer(cartReducer, initialState);
  const { isLoading, error, isAdding } = state;
  const { data: productsFromSwr } = useCartProducts();

  const handleAddToCart = useCallback(async (product: Product): Promise<void> => {
    try {
      dispatch({ type: ADD_PRODUCT_REQUEST });

      const data = await addToCart(product);

      mutate(CART_API);

      dispatch({ type: ADD_PRODUCT_SUCCESS, payload: data });
    } catch {
      dispatch({ type: ADD_PRODUCT_FAILURE });
    }
  }, []);

  const handleDeleteProductFromCart = useCallback(async (productId: string): Promise<void> => {
    try {
      dispatch({ type: DELETE_PRODUCT_REQUEST });

      await deleteCartItem(productId);

      mutate(CART_API);

      dispatch({ type: DELETE_PRODUCT_SUCCESS, payload: productId });
    } catch (error) {
      dispatch({ type: DELETE_PRODUCT_FAILURE });
    }
  }, []);

  const handleUpdateItem = useCallback(
    async (productId: string, updatedData: Partial<Product>, url: string) => {
      try {
        dispatch({ type: UPDATE_PRODUCT_REQUEST });
        await updateItem(productId, updatedData, url);

        mutate(CART_API);

        const updatedProductData: Partial<Product> = {
          ...updatedData,
          id: productId,
        };

        dispatch({ type: UPDATE_PRODUCT_SUCCESS, payload: updatedProductData });
      } catch {
        dispatch({ type: UPDATE_PRODUCT_FAILURE });
      }
    },
    []
  );

  const cartContextValue = useMemo(
    () => ({
      isLoading,
      error,
      products: productsFromSwr || [],
      isAdding,
      onAddToCart: handleAddToCart,
      onUpdateCart: handleUpdateItem,
      onDeleteProductFromCart: handleDeleteProductFromCart,
    }),
    [state, productsFromSwr]
  );

  return <CartContext.Provider value={cartContextValue}>{children}</CartContext.Provider>;
};

export default CartProvider;
