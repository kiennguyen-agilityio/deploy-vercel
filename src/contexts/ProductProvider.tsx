import { createContext, ReactNode, useState, useMemo } from 'react';

// hooks
import { useProducts } from '@/components/hooks/fetch';

// interfaces
import Product from '@/interface/Product';

interface ProductProviderProps {
  children: ReactNode;
}

export interface ListProductContextProps {
  products?: Product[];
  selectedProduct?: Product | null;
  onGetProductById: (productId: string) => void;
  error: string;
  isLoading: boolean;
}

export const ListProductContext = createContext<ListProductContextProps>({
  onGetProductById: async () => {},
  error: '',
  isLoading: false,
});

export const ListProductProvider: React.FC<ProductProviderProps> = ({ children }) => {
  const { data: productsFromSwr, error: productsError, isLoading } = useProducts();
  const [selectedProduct, setSelectedProduct] = useState<Product | null>(null);

  const getProductById = (productId: string) => {
    const selectedProduct = productsFromSwr?.find((product) => product.id === productId);
    setSelectedProduct(selectedProduct || null);
  };

  const productContextValue = useMemo(
    () => ({
      products: productsFromSwr,
      selectedProduct,
      onGetProductById: getProductById,
      error: productsError ? productsError.message : '',
      isLoading,
    }),
    [productsError, productsFromSwr, selectedProduct]
  );

  return (
    <ListProductContext.Provider value={productContextValue}>
      {children}
    </ListProductContext.Provider>
  );
};
