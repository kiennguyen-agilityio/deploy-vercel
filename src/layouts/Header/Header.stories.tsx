import { StoryObj, Meta } from '@storybook/react';
import { MemoryRouter } from 'react-router-dom';
import Header from './Header';

export default {
  title: 'Components/Header',
  component: Header,
  decorators: [
    (Story) => (
      <MemoryRouter>
        <Story />
      </MemoryRouter>
    ),
  ],
} as Meta;

type Story = StoryObj<typeof Header>;

export const Default: Story = {};
