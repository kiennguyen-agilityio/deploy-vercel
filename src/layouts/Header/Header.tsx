import { useContext, memo } from 'react';
import { Link } from 'react-router-dom';

// contexts
import { SearchContext } from '@/contexts/SearchProvider';

// components
import SearchBar from '@/components/SearchBar/SearchBar';
import Logo from './Logo';

// style
import './Header.css';
import { useCartProducts } from '@/components/hooks/fetch';

const Header: React.FC = () => {
  const { data: products } = useCartProducts();

  const { searchValue } = useContext(SearchContext);
  const { onSearchValue } = useContext(SearchContext);

  return (
    <div className="navbar">
      <Logo />

      <SearchBar searchValue={searchValue} onSetSearchValue={onSearchValue} />

      <div className="icons">
        <Link to="/cart" className="cart-icon-container">
          <i className="material-symbols-outlined icon">shopping_cart</i>
          {products?.length && <span className="cart-item-count">{products?.length}</span>}
        </Link>
        <i className="material-symbols-outlined icon">favorite</i>
      </div>
      <div className="button-container">
        <button className="btn-sign-up" type="button">
          Sign Up
        </button>
        <button className="btn-login" type="button">
          Login
        </button>
      </div>
    </div>
  );
};

export default memo(Header);
