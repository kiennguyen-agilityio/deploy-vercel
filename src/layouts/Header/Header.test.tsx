import { render, fireEvent } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
// contexts
import { CartContext, CartContextProps } from '@/contexts/CartProvider';
import { SearchContext, SearchContextProps } from '@/contexts/SearchProvider';

// components
import Header from './Header';
import Product from '@/interface/Product';

describe('Header component', () => {
  const defaultProps = {
    cart: [
      {
        id: 1,
        name: 'Fullset Black Chair & Sofa',
        sale: '350',
        price: 5000,
        discountPercent: 10,
        image: '../src/assets/images/full-black-chair-sofa.png',
        descriptions:
          'black sofa fullset, made of leather selection from the Himalayan mountains. Knitted by the professional hands of housewives. Making a masterpiece of the best sofa in the world. Very soft and comfortable',
        quantity: 2,
        color: 'Red',
        subcategory: 'Office chair',
      },
    ],
    onSearchValue: jest.fn(),
    onResetSearchValue: jest.fn(),
  };

  interface Props {
    cart: Product[];
    onSearchValue: () => void;
    onResetSearchValue: () => void;
  }

  const setup = (props: Props = defaultProps) => {
    const cartContextValue: CartContextProps = {
      cart: props.cart || defaultProps.cart,
      addToCart: jest.fn(),
      updateProductInCart: jest.fn(),
      updateProductDetail: jest.fn(),
      deleteProductFromCart: jest.fn(),
    };

    const searchContextValue: SearchContextProps = {
      searchValue: '',
      onSearchValue: props.onSearchValue || defaultProps.onSearchValue,
      onResetSearchValue: props.onResetSearchValue || defaultProps.onResetSearchValue,
    };

    return render(
      <BrowserRouter basename="/">
        <CartContext.Provider value={cartContextValue}>
          <SearchContext.Provider value={searchContextValue}>
            <Header />
          </SearchContext.Provider>
        </CartContext.Provider>
      </BrowserRouter>
    );
  };

  it('renders correctly', () => {
    const { container } = setup();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('displays the correct number of items in the cart', () => {
    const { getByText } = setup();
    const cartItemCount = getByText((content, element) => {
      if (!element) return false;
      const classNames = element.getAttribute('className');
      if (classNames && classNames.includes('cart-item-count')) {
        return true;
      }
      const normalizedContent = content.trim();
      if (normalizedContent === '1') {
        return true;
      }
      return false;
    });
    expect(cartItemCount).toBeInTheDocument();
  });

  it('calls the onSearchValue function with the correct value when typing in the search input', () => {
    const { getByRole } = setup();
    const searchInput = getByRole('textbox');
    const searchValue = 'Product';
    fireEvent.change(searchInput, { target: { value: searchValue } });
    expect(defaultProps.onSearchValue).toHaveBeenCalledWith(searchValue);
  });
});
