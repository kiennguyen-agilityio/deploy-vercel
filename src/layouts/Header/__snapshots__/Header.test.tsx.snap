// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component renders correctly 1`] = `
<div
  class="navbar"
>
  <a
    href="/"
  >
    <div
      class="logo"
    >
      <i
        class="fa-sharp fa-solid fa-bag-shopping"
      />
      <span
        class="shop-name"
      >
        Shop Bags
      </span>
    </div>
  </a>
  <div
    class="search-bar-container"
  >
    <input
      class="search-bar"
      value=""
    />
    <i
      class="search-icon fa-sharp fa-solid fa-search"
    />
  </div>
  <div
    class="icons"
  >
    <a
      class="cart-icon-container"
      href="/cart"
    >
      <i
        class="material-symbols-outlined icon"
      >
        shopping_cart
      </i>
      <span
        class="cart-item-count"
      >
        1
      </span>
    </a>
    <i
      class="material-symbols-outlined icon"
    >
      favorite
    </i>
  </div>
  <div
    class="button-container"
  >
    <button
      class="btn-sign-up"
      type="button"
    >
      Sign Up
    </button>
    <button
      class="btn-login"
      type="button"
    >
      Login
    </button>
  </div>
</div>
`;

exports[`Header component should render correctly 1`] = `
<DocumentFragment>
  <div
    class="navbar"
  >
    <a
      href="/"
    >
      <div
        class="logo"
      >
        <i
          class="fa-sharp fa-solid fa-bag-shopping"
        />
        <span
          class="shop-name"
        >
          Shop Bags
        </span>
      </div>
    </a>
    <div
      class="search-bar-container"
    >
      <input
        class="search-bar"
        value=""
      />
      <i
        class="search-icon fa-sharp fa-solid fa-search"
      />
    </div>
    <div
      class="icons"
    >
      <a
        class="cart-icon-container"
        href="/cart"
      >
        <i
          class="material-symbols-outlined icon"
        >
          shopping_cart
        </i>
      </a>
      <i
        class="material-symbols-outlined icon"
      >
        favorite
      </i>
    </div>
    <div
      class="button-container"
    >
      <button
        class="btn-sign-up"
        type="button"
      >
        Sign Up
      </button>
      <button
        class="btn-login"
        type="button"
      >
        Login
      </button>
    </div>
  </div>
</DocumentFragment>
`;
